### ISSUE TITLE
`unused_import` false positives

### ISSUE THREAD

**Issue by AdrianBinDC** [realm/SwiftLint#3116]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I ran `swiftlint analyze` on a large project with autocorrect enabled for `unused_imports`. While swiftlint found a large amount of unused imports, the project didn't compile when I built after running `swiftlint analyze`. I stashed swiftlint's changes before making alterations to it. From there, I applied the stash and kept addressing compiler errors until I got the project to build. When I finished, I discovered despite the reduction in imports, the project took slightly longer to build when building after nuking derived data. When building WITH derived data, the build time was dramatically faster before removing imports than after.

Should I expect compile times to go down as a result of removing unused imports?

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
 Done linting! Found X violations, 0 serious in Y files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
0.39.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
Homebrew
* Which Xcode version are you using (check `xcodebuild -version`)?
Xcode 11.3.1
Build version 11C504

**Comment by jpsim** [comment]
You seem to be discussing two separate points, first that correcting the unused import rule makes your modules fail to compile, second that after fixing those missing imports your module takes longer to compile. I’ll separate these discussion points for clarity.

### 1. Modules no longer compile after correcting for unused imports

This is because Swift allows referencing some declarations that are only transitively imported, but the rule will remove all imports that aren’t _directly_ used, even if they transitively expose a module that _is_ used in that file. Up until now, re-adding these directly used imports has required manual editing after running SwiftLint Analyze.

I’m in the process of adding configuration options to the unused import rule in https://github.com/realm/SwiftLint/pull/3117 that would require that all modules referenced in a source file be _explicitly_ imported in that file. This adds a lot of noise to the import list, but has the added advantage of preventing `swiftlint lint --autocorrect` from introducing compiler errors in your files.

I’d love to know if you’d be interested in using that feature, or if you have suggestions for how it should work.

### 2. Adding new imports after correcting unused import violations makes compilation take longer

I was surprised to read this considering that the imports you’re adding are presumably direct imports for the exact modules referenced in the source file. I’d need to see a sample project with steps to reproduce to be able to offer more advice here. It may also be a swift compiler issue if narrowing the imports included in a file lengthens its compilation time.
